crates/rattler_repodata_gateway/src/gateway/query.rs

```diff
@@ -247,8 +247,15 @@ impl GatewayQuery {
                     // Add the records to the result
                     if records.len() > 0 {
                         let result = &mut result[result_idx];
-                        result.len += records.len();
-                        result.shards.push(records);
+
+                        for record in records.iter() {
+                            if !self.recursive && !request_specs.iter().any(|spec| spec.matches(record)) {
```

Comment:
`!self.recursive` is a hack... I don't know how we should apply filters when doing recursive queries. Does anyone have an idea of what we should do?

Reply:
I guess what I would need to do is to only filter "top level" packages when in recursive mode and ignore specs that are direct URL specs (since direct URLs are already correctly filtered?

Reply:
Yes! But that will already happen correctly because the specs added for the dependencies match all records.

Regardless I think this is better because now we also don't pay for the matches check if the call is not recursive anyway.